python/global_search_completer/global_search_completer.py

```diff
@@ -0,0 +1,362 @@
+
+    def set_bg_task_manager(self, task_manager):
+        """
+        Specify the background task manager to use to pull
+        data in the background. Data calls
+        to Shotgun will be dispatched via this object.
+
+        :param task_manager: Background task manager to use
+        :type  task_manager: :class:`~tk-framework-shotgunutils:task_manager.BackgroundTaskManager`
+        """
+        self.__sg_data_retriever = shotgun_data.ShotgunDataRetriever(
+            self, bg_task_manager=task_manager)
+
+        self.__sg_data_retriever.start()
+        self.__sg_data_retriever.work_completed.connect(self.__on_worker_signal)
+        self.__sg_data_retriever.work_failure.connect(self.__on_worker_failure)
+
+    def set_searchable_entity_types(self, types_dict):
+        """
+        Specify a dictionary of entity types with optional search filters to
+        limit the breadth of the widget's search.
+
+        Use this method to override the default searchable entity types
+        dictionary which looks like this::
+
+          {
+            "Asset": [],
+            "Shot": [],
+            "Task": [],
+            "HumanUser": [["sg_status_list", "is", "act"]],    # only active users
+            "Group": [],
+            "ClientUser": [["sg_status_list", "is", "act"]],   # only active users
+            "ApiUser": [],
+            "Version": [],
+            "PublishedFile": [],
+          }
+
+        :param types_dict: A dictionary of searchable types with optional filters
+
+        """
+        self._entity_search_criteria = types_dict
+
+
+    ############################################################################
+    # internal methods
+
+    def _clear_model(self, add_loading_item=True):
+        """
+        Clears the current data in the model.
+
+        :param add_loading_item: if true, a "loading please wait" item will
+                                 be added.
+        """
+        # clear model
+        self.model().clear()
+
+        if add_loading_item:
+            item = QtGui.QStandardItem("Loading data...")
+            item.setData(self.MODE_LOADING, self.MODE_ROLE)
+            self.model().appendRow(item)
+
+    def _do_sg_global_search(self, sg, data):
+        """
+        Actual payload for running a global search in Shotgun.
+
+        Note: This runs in a different thread and cannot access
+        any QT UI components. It should not do any logging and should
+        be as minimal as possible.
+
+        :param sg: Shotgun instance
+        :param data: data dictionary passed in from _submit(). This dict
+            carries the options payload and is forwarded over by the
+            sg_data.execute_method() method. In our case, the dictionary
+            holds a single 'text' key which contains the search phrase.
+        """
+
+        # constrain by project in the search
+        project_ids = []
+        if self._bundle.context.project:
```

Comment:
Conceptually, Context represents a location in Shotgun, whereas pipeline_configuration.get_project_id() represents the project that the current project is associated with. In the case of any shotgun related stuff like this, use context. 

I think right now, they will always be the same, but if we wanted to allow for more cross project action going on in the future, we would be better set up with this being context based (as it is now).